coder_sniffer/Drupal/Sniffs/Commenting/DeprecatedSniff.php

```diff
@@ -75,20 +74,72 @@ public function process(File $phpcsFile, $stackPtr)
             }
         }
 
-        $depText = trim($depText);
-
         // The standard format for the deprecation text is:
-        // @deprecated in %in-version% and will be removed from %removal-version%. %extra-info%.
-        // Use (?U) 'ungreedy' before the version so that only the text up to
-        // the first '. ' is matched, as there may be more than one sentence in
-        // the extra-info part.
-        $matches = [];
-        preg_match('/in (.+) and is removed from (?U)(.+)\. (.+)$/', $depText, $matches);
+        // @deprecated in %in-version% and is removed from %removal-version%. %extra-info%.
+        $standardFormat = "@deprecated in %%deprecation-version%% and is removed from %%removal-version%%. %%extra-info%%.";
+
+        // Use (?U) 'ungreedy' before the removal-version so that only the text
+        // up to the first dot+space is matched, as there may be more than one
+        // sentence in the extra-info part.
+        $fullText = trim(implode(' ', $textItems));
+        $matches  = [];
+        preg_match('/^in (.+) and is removed from (?U)(.+)(?:\. | |\.$|$)(.*)$/', $fullText, $matches);
         // There should be 4 items in $matches: 0 is full text, 1 = in-version,
-        // 2 = removal-version, 3 = extra-info.
+        // 2 = removal-version, 3 = extra-info (can be blank at this stage).
         if (count($matches) !== 4) {
-            $error = "The text '@deprecated %s' does not match the standard format: @deprecated in %%deprecation-version%% and is removed from %%removal-version%%. %%extra-info%%.";
-            $phpcsFile->addError($error, $stackPtr, 'IncorrectTextLayout', [$depText]);
+            // The full text does not match the standard. Try to find fixes by
+            // testing with a relaxed set of criteria, based on common
+            // formatting variations. This is designed for Core fixes.
+            $error = "The text '@deprecated %s' does not match the standard format: ".$standardFormat;
+            // All of the standard text should be on the first comment line, so
+            // try to match with common formatting errors to allow an automatic
+            // fix. If not possible then report a normal error.
+            $matchesFix = [];
+            $fix        = null;
+            if (count($textItems) > 0) {
+                // Get just the first line of the text.
+                $key   = array_keys($textItems)[0];
+                $text1 = $textItems[$key];
+                preg_match('/^(.*)(as of|in) (drupal|)( |:|)+([\d\.\-xdev\?]+)(,| |. |)(.*)(removed|removal)([ |from|before|in|the]*) (drupal|)( |:|)([\d\-\.xdev]+)( |,|$)+(?:release|)(?:[\.,])*(.*)$/i', $text1, $matchesFix);
```

Comment:
we should not hard-code "drupal" here, right?

Reply:
Thanks for all the review comments. Very helpful. I will start with this one, as it affects how I proceed with the others. If we do not hard-code 'drupal' here then it would mean catering for contrib within the fixable section. I guess that preg_match could be altered to allow any word+undercore instead of 'drupal' but that would mean later testing to see if it was 'drupal' or blank (in which case the fix is for core) or some other word (in which case the fix is for contrib). The version checking would have to be different for contrib, but that could be done. 

If we think this is worth it, I will try to make the changes. But I'm not sure how much deprecation there is in contrib, so I was happy initially to exclude it from automatic fixing. 

The hard-coding of 'drupal' at that line effectively says "I am only attempting to provide auto-fixes for core problems". If I made that clearer in the comments would that be an acceptable solution? I think the alternative (if you do not want 'drupal' hard-coded there) would be for  me to re-write the whole patch to remove the auto-fixable part completely. But it does need the new sniff to cater for missing 'extra info' so there would be some work to do.

Reply:
Yep, hard coding "drupal" just for the fixable error as first step would be acceptable I think. Please add code comments to explain that.

Thank you!